streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableKTableJoinMerger.java

```diff
@@ -21,29 +21,66 @@
 import org.apache.kafka.streams.processor.ProcessorContext;
 import org.apache.kafka.streams.state.KeyValueStore;
 
+import java.util.ArrayList;
+
 class KTableKTableJoinMerger<K, V> implements KTableProcessorSupplier<K, V, V> {
 
     private final KTableImpl<K, ?, V> parent1;
     private final KTableImpl<K, ?, V> parent2;
     private final String queryableName;
     private boolean sendOldValues = false;
 
-    public KTableKTableJoinMerger(final KTableImpl<K, ?, V> parent1,
-                                  final KTableImpl<K, ?, V> parent2,
-                                  final String queryableName) {
+    KTableKTableJoinMerger(final KTableImpl<K, ?, V> parent1,
+                           final KTableImpl<K, ?, V> parent2,
+                           final String queryableName) {
         this.parent1 = parent1;
         this.parent2 = parent2;
         this.queryableName = queryableName;
     }
 
     @Override
     public Processor<K, Change<V>> get() {
-        return new KTableKTableJoinMergeProcessor<>();
+        return new KTableKTableJoinMergeProcessor();
     }
 
     @Override
     public KTableValueGetterSupplier<K, V> view() {
-        return parent1.valueGetterSupplier();
+        // if the result KTable is materialized, use the materialized store to return getter value;
+        // otherwise rely on the parent getter and apply join on-the-fly
+        if (queryableName != null) {
+            return new KTableValueGetterSupplier<K, V>() {
+                public KTableValueGetter<K, V> get() {
+                    return new KTableMaterializedValueGetter<>(queryableName);
+                }
+
+                @Override
+                public String[] storeNames() {
+                    return new String[]{queryableName};
+                }
+            };
+        } else {
+            return new KTableValueGetterSupplier<K, V>() {
+
+                public KTableValueGetter<K, V> get() {
+                    return parent1.valueGetterSupplier().get();
+                }
+
+                @Override
+                public String[] storeNames() {
+                    // we need to allow the downstream processor to be able to access both ends of the joining table's value getters
+                    final String[] storeNames1 = parent1.valueGetterSupplier().storeNames();
+                    final String[] storeNames2 = parent2.valueGetterSupplier().storeNames();
+                    final ArrayList<String> stores = new ArrayList<>(storeNames1.length + storeNames2.length);
```

Comment:
Why don't using an array from the beginning on?

Reply:
This is to be coherent with `Topology#connectProcessorAndStateStores` with array store names.

Reply:
Cannot follow. We need to return an array be create a ArryaList that we convert in an array later. Why not an array directly instead of an ArrayList?

Reply:
`storeNames` is used in `connectProcessorAndStateStores`, which can also be called by user to connect one processor with multiple state stores, where state store names can be passed as a enum array.

Reply:
I understand that. I am just wondering, why we create a `ArrayList` here in stead of a plain array:
```
final String[] stores = new String[storeNames1.length + storeNames2.length]
int i = 0;
for (int j = 0; j < storeNames1.length; ++j, ++i) {
    stores[i] = storeNames1[j]
}
for (int j = 0; j < storeNames2.length; ++j, ++i) {
    stores[i] = storeNames2[j]
}
return stores;
```

Reply:
Ah I see what you mean, ack.